File: src/lib/bridge/research.ts
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type {
  HarnessConfiguration,
  HarnessConfigurationVersion,
  HarnessChangeSet,
  EffectiveInstructionStack,
  HarnessRun,
  HarnessSnapshot,
  Search,
  SearchCandidate,
  SearchCandidatesPreview,
  SearchDraft,
  SearchUpdated,
  RunReconciliationPlan,
  ResearchCheckpoint,
} from "$lib/domain/research";
import type { ResearchStateSnapshot } from "$lib/domain/research-state";

export async function createSearch(draft: SearchDraft): Promise<Search> {
  return invoke<Search>("create_search", { draft });
}

export async function listSearches(): Promise<Search[]> {
  return invoke<Search[]>("list_searches");
}

export async function getSearch(searchId: string): Promise<Search> {
  return invoke<Search>("get_search", { searchId });
}

export async function listSearchCandidates(searchId: string): Promise<SearchCandidate[]> {
  return invoke<SearchCandidate[]>("list_search_candidates", { searchId });
}

/** Enqueue a run; resolves with the run id. Progress arrives via events. */
export async function runSearch(searchId: string, mode = "deep"): Promise<string> {
  return invoke<string>("run_search", { searchId, mode });
}

export async function cancelSearchRun(runId: string): Promise<void> {
  return invoke<void>("cancel_search_run", { runId });
}

export async function getResearchHarness(projectId: string): Promise<HarnessSnapshot> {
  return invoke<HarnessSnapshot>("get_research_harness", { projectId });
}

export async function saveResearchHarness(
  projectId: string,
  configuration: HarnessConfiguration,
): Promise<HarnessSnapshot> {
  return invoke<HarnessSnapshot>("save_research_harness", { projectId, configuration });
}

export async function listHarnessConfigurationVersions(
  projectId: string,
): Promise<HarnessConfigurationVersion[]> {
  return invoke<HarnessConfigurationVersion[]>("list_harness_configuration_versions", {
    projectId,
  });
}

export async function clearHarnessConfigurationHistory(projectId: string): Promise<number> {
  return invoke<number>("clear_harness_configuration_history", { projectId });
}

export async function getHarnessRunInstructions(
  runId: string,
): Promise<EffectiveInstructionStack> {
  return invoke<EffectiveInstructionStack>("get_harness_run_instructions", { runId });
}

export async function getResearchCheckpoint(runId: string): Promise<ResearchCheckpoint> {
  return invoke<ResearchCheckpoint>("get_research_checkpoint", { runId });
}

export async function listResearchCheckpoints(projectId: string): Promise<ResearchCheckpoint[]> {
  return invoke<ResearchCheckpoint[]>("list_research_checkpoints", { projectId });
}

export async function restoreResearchCheckpoint(
  projectId: string,
  runId: string,
  expectedCurrentRevision: number,
): Promise<ResearchStateSnapshot> {
  return invoke<ResearchStateSnapshot>("restore_research_checkpoint", {
    projectId,
    runId,
    expectedCurrentRevision,
  });
}

export async function getHarnessChangeSet(runId: string): Promise<HarnessChangeSet> {
  return invoke<HarnessChangeSet>("get_harness_change_set", { runId });
}

export async function applyHarnessChangeSet(id: string): Promise<HarnessChangeSet> {
  return invoke<HarnessChangeSet>("apply_harness_change_set", { id });
}

export async function rejectHarnessChangeSet(
  id: string,
  reason: string,
): Promise<HarnessChangeSet> {
  return invoke<HarnessChangeSet>("reject_harness_change_set", { id, reason });
}

export async function editHarnessChangeSet(
  id: string,
  plan: RunReconciliationPlan,
): Promise<HarnessChangeSet> {
  return invoke<HarnessChangeSet>("edit_harness_change_set", { id, patch: { plan } });
}

export async function runProjectResearch(projectId: string): Promise<HarnessRun> {
  return invoke<HarnessRun>("run_project_research", { projectId });
}

export async function cancelProjectResearch(projectId: string): Promise<void> {
  return invoke<void>("cancel_project_research", { projectId });
}

export async function pauseResearchHarness(projectId: string): Promise<HarnessSnapshot> {
  return invoke<HarnessSnapshot>("pause_research_harness", { projectId });
}

export async function resumeResearchHarness(projectId: string): Promise<HarnessSnapshot> {
  return invoke<HarnessSnapshot>("resume_research_harness", { projectId });
}

export async function stopResearchHarness(projectId: string): Promise<HarnessSnapshot> {
  return invoke<HarnessSnapshot>("stop_research_harness", { projectId });
}

export async function markSearchCandidatesSeen(searchId: string): Promise<void> {
  return invoke<void>("mark_search_candidates_seen", { searchId });
}

export async function markSearchCandidateSaved(
  candidateId: string,
  saved: boolean,
): Promise<void> {
  return invoke<void>("mark_search_candidate_saved", { candidateId, saved });
}

/** Subscribe to background run progress. Returns an unlisten function. */
export async function listenSearchUpdated(
  onEvent: (payload: SearchUpdated) => void,
): Promise<UnlistenFn> {
  return listen<SearchUpdated>("search_updated", (event) => onEvent(event.payload));
}

/** Subscribe to transient deep-search candidate previews. */
export async function listenSearchCandidatesPreview(
  onEvent: (payload: SearchCandidatesPreview) => void,
): Promise<UnlistenFn> {
  return listen<SearchCandidatesPreview>("search_candidates_preview", (event) =>
    onEvent(event.payload),
  );
}

File: src-tauri/src/commands/research.rs
//! Tauri commands for deep-research agentic search (RFC 0037).

use crate::domain::harness::{
    EffectiveInstructionStack, HarnessConfiguration, HarnessConfigurationVersion, HarnessRun,
    HarnessRunTrigger, HarnessSnapshot, ResearchCheckpoint,
};
use crate::domain::reconciliation::{HarnessChangeSet, HarnessChangeSetPatch};
use crate::domain::research::{Search, SearchCandidate, SearchDraft};
use crate::domain::research_state::ResearchStateSnapshot;
use crate::services::research::controller::ProjectResearchController;
use crate::services::research::manager::SearchManager;
use crate::storage::library_store::LibraryStore;

#[tauri::command]
pub fn create_search(
    store: tauri::State<'_, LibraryStore>,
    draft: SearchDraft,
) -> Result<Search, String> {
    store.create_search(&draft)
}

#[tauri::command]
pub fn list_searches(store: tauri::State<'_, LibraryStore>) -> Result<Vec<Search>, String> {
    store.list_searches()
}

#[tauri::command]
pub fn get_search(
    store: tauri::State<'_, LibraryStore>,
    search_id: String,
) -> Result<Search, String> {
    store.get_search(&search_id)
}

#[tauri::command]
pub fn list_search_candidates(
    store: tauri::State<'_, LibraryStore>,
    search_id: String,
) -> Result<Vec<SearchCandidate>, String> {
    store.list_search_candidates(&search_id)
}

/// Enqueue a run (manual or re-run) and return its run id immediately. The UI
/// then follows progress via `search_updated` events.
#[tauri::command]
pub fn run_search(
    manager: tauri::State<'_, SearchManager>,
    search_id: String,
    mode: Option<String>,
) -> Result<String, String> {
    manager.run_search(search_id, mode)
}

#[tauri::command]
pub fn cancel_search_run(manager: tauri::State<'_, SearchManager>, run_id: String) {
    manager.cancel_run(&run_id);
}

#[tauri::command]
pub fn mark_search_candidate_saved(
    store: tauri::State<'_, LibraryStore>,
    candidate_id: String,
    saved: bool,
) -> Result<(), String> {
    store.mark_search_candidate_saved(&candidate_id, saved)
}

#[tauri::command]
pub fn mark_search_candidates_seen(
    store: tauri::State<'_, LibraryStore>,
    search_id: String,
) -> Result<(), String> {
    store.mark_search_candidates_seen(&search_id)
}

#[tauri::command]
pub fn get_research_harness(
    store: tauri::State<'_, LibraryStore>,
    project_id: String,
) -> Result<HarnessSnapshot, String> {
    store.get_harness_snapshot(&project_id)
}

#[tauri::command]
pub fn save_research_harness(
    store: tauri::State<'_, LibraryStore>,
    project_id: String,
    configuration: HarnessConfiguration,
) -> Result<HarnessSnapshot, String> {
    store.save_harness_configuration(&project_id, &configuration)
}

#[tauri::command]
pub fn list_harness_configuration_versions(
    store: tauri::State<'_, LibraryStore>,
    project_id: String,
) -> Result<Vec<HarnessConfigurationVersion>, String> {
    store.list_harness_configuration_versions(&project_id)
}

#[tauri::command]
pub fn clear_harness_configuration_history(
    store: tauri::State<'_, LibraryStore>,
    project_id: String,
) -> Result<usize, String> {
    store.clear_harness_configuration_history(&project_id)
}

#[tauri::command]
pub fn get_harness_run_instructions(
    store: tauri::State<'_, LibraryStore>,
    run_id: String,
) -> Result<EffectiveInstructionStack, String> {
    store.get_harness_run_instructions(&run_id)
}

#[tauri::command]
/// Loads one persisted Run checkpoint.
pub fn get_research_checkpoint(
    store: tauri::State<'_, LibraryStore>,
    run_id: String,
) -> Result<ResearchCheckpoint, String> {
    store.get_research_checkpoint(&run_id)
}

#[tauri::command]
/// Lists persisted Run checkpoints for a Project.
pub fn list_research_checkpoints(
    store: tauri::State<'_, LibraryStore>,
    project_id: String,
) -> Result<Vec<ResearchCheckpoint>, String> {
    store.list_research_checkpoints(&project_id)
}

#[tauri::command]
/// Appends a new State revision matching the selected same-Project checkpoint.
pub fn restore_research_checkpoint(
    store: tauri::State<'_, LibraryStore>,
    project_id: String,
    run_id: String,
    expected_current_revision: i64,
) -> Result<ResearchStateSnapshot, String> {
    store.restore_research_checkpoint(&project_id, &run_id, expected_current_revision)
}

#[tauri::command]
/// Loads the immutable Project-change record associated with a completed Run.
pub fn get_harness_change_set(
    store: tauri::State<'_, LibraryStore>,
    run_id: String,
) -> Result<HarnessChangeSet, String> {
    store.get_harness_change_set(&run_id)
}

#[tauri::command]
/// Applies a proposed Change Set atomically within the Run's authority snapshot.
pub fn apply_harness_change_set(
    store: tauri::State<'_, LibraryStore>,
    id: String,
) -> Result<HarnessChangeSet, String> {
    store.apply_harness_change_set(&id)
}

#[tauri::command]
/// Retains but rejects a proposed Change Set with a researcher-supplied reason.
pub fn reject_harness_change_set(
    store: tauri::State<'_, LibraryStore>,
    id: String,
    reason: String,
) -> Result<HarnessChangeSet, String> {
    store.reject_harness_change_set(&id, &reason)
}

#[tauri::command]
/// Replaces bounded editable plan fields after full server-side revalidation.
pub fn edit_harness_change_set(
    store: tauri::State<'_, LibraryStore>,
    id: String,
    patch: HarnessChangeSetPatch,
) -> Result<HarnessChangeSet, String> {
    store.edit_harness_change_set(&id, &patch.plan)
}

#[tauri::command]
pub fn run_project_research(
    controller: tauri::State<'_, ProjectResearchController>,
    project_id: String,
) -> Result<HarnessRun, String> {
    controller.start(&project_id, HarnessRunTrigger::Manual, None)
}

#[tauri::command]
pub fn pause_research_harness(
    store: tauri::State<'_, LibraryStore>,
    project_id: String,
) -> Result<HarnessSnapshot, String> {
    store.pause_research_harness(&project_id)
}

#[tauri::command]
pub fn resume_research_harness(
    store: tauri::State<'_, LibraryStore>,
    project_id: String,
) -> Result<HarnessSnapshot, String> {
    store.resume_research_harness(&project_id)
}

#[tauri::command]
pub async fn stop_research_harness(
    store: tauri::State<'_, LibraryStore>,
    controller: tauri::State<'_, ProjectResearchController>,
    project_id: String,
) -> Result<HarnessSnapshot, String> {
    let snapshot = store.stop_research_harness(&project_id)?;
    if snapshot.runs.iter().any(|run| {
        matches!(
            run.status.as_str(),
            "queued" | "planning" | "searching" | "assessing" | "ranking" | "reconciling"
        )
    }) {
        controller.cancel(&project_id).await?;
    }
    Ok(snapshot)
}

#[tauri::command]
pub async fn cancel_project_research(
    controller: tauri::State<'_, ProjectResearchController>,
    project_id: String,
) -> Result<(), String> {
    controller.cancel(&project_id).await
}
